### ISSUE TITLE
get! corrupts dictionary if modified in function

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#7944]
```julia
julia> a = Dict()
Dict{Any,Any} with 0 entries

julia> get!(a, 1) do
       get!(a, 2) do
       get!(a, 3) do
       get!(a, 4) do
       get!(a, 5) do
       get!(a, 6) do
       get!(a, 7) do
       get!(a, 8) do
       get!(a, 9) do
       get!(a, 10) do
       10
       end; 9
       end; 8
       end; 7
       end; 6
       end; 5
       end; 4
       end; 3
       end; 2
       end; 1
       end
1

julia> a
Dict{Any,Any} with 10 entries:
  7  => 7
  4  => 4
  10 => 10
  2  => 2
  3  => 3
  5  => 5
  6  => 6
  1  => 1
```

...what happened to the other 2 values?

Not sure if this can be made to work without introducing additional overhead, but if not it should be documented. I almost wrote code that did this.

**Comment by garrison** [comment]
I just got bit by this one.  I've been using this in a piece of code for several months, and just discovered today some slight inconsistencies.  What I really don't like about it is that it leads to code that *mostly* works, all while corrupting an internal data structure.

I like the approach of having a `dirty` bit, as implemented in #9595.

**Comment by ScottPJones** [comment]
Any reason why #9595 hasn't been merged in yet?

**Comment by garrison** [comment]
Well, @JeffBezanson made it an RFC, and nobody commented directly there until now :).  At the moment there are merge conflicts to resolve, but I think it seems like a very nice and reasonable solution.

**Comment by ScottPJones** [comment]
Yes, it looked good to me

**Comment by JeffBezanson** [comment]
Fixed by #9595.
